### PR TITLE
Add NSNotification testing features

### PR DESCRIPTION
**Pull request by bgerstle**
- [x] Test 1+ posted notification matches success
- [x] Test 0 posted notification success
- [x] Test notification name match failure
- [x] Test notification object match failure
- [x] Test eventual notification match
- [ ] Add ObjC support
- [ ] Update README

---

I was writing some tests for an object that uses `NSNotification` to broadcast will/did state change events globally.  To test this, I needed to verify whether a notification was or wasn't posted.  This was partially doable with XCTest's `expectationForNotification`, but I figured it would be cooler—I mean more expressive/readable 🤓—to use a Nimble matcher.

### CODE REVIEW

File: Nimble/Utils/NotificationCollector.swift [hunk 1]
```diff
@@ -0,0 +1,50 @@
+//
+//  NotificationCollector.swift
+//  Nimble
+//
+//  Created by Brian Gerstle on 12/31/15.
+//  Copyright © 2015 Jeff Hui. All rights reserved.
+//
```
Comment (jeffh) [inline Nimble/Utils/NotificationCollector.swift]: Nitpick: Please remove the header comments. Mostly just to be consistent with the rest of the code base (and they don't add much value)

File: Nimble/Utils/NotificationCollector.swift [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    var once: Bool = false
+    return {
+        assert(NSThread.isMainThread(), "Only expecting closure to be evaluated on main thread.")
+        if !once {
+            once = true
+            try block()
+        }
+        return collector.observedNotifications
+    }
+}
```
Comment (jeffh) [inline Nimble/Utils/NotificationCollector.swift]: I think the `postedNotifications` func is causing problems that forces `expect(expression: ...)` form you're talking about. `postedNotification` should return the observed notifications:

```swift
public func postedNotifications(notificationCenter: NSNotificationCenter = NSNotificationCenter.defaultCenter(),
                                block: () throws -> Void) throws -> [NSNotification]
```

[Nimble's expect(...)](https://github.com/Quick/Nimble/blob/master/Nimble/DSL.swift#L2-L17) argument is an [@autoclosure](https://developer.apple.com/library/ios/documentation/Swift/Conceptual/Swift_Programming_Language/Attributes.html). This means the expression in that argument is an implicit closure:

```swift
// this line
expect(postedNotification { ... }).to(equal(...))
// is equivalent to this line
expect { postedNotification { ... } }.to(equal(...))
```

This allows Nimble to catch errors despite looking like a normal argument:

```swift
expect(try somethingThatThrows()).to(throwError(...))
expect(raisesNSException()).to(raiseException(...))
```

File: NimbleTests/Matchers/PostNotificationTest.swift [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+            NSNotificationCenter.defaultCenter().postNotification(n1)
+            NSNotificationCenter.defaultCenter().postNotification(n2)
+        }).to(equal([n1, n2]))
+    }
+
+// MARK: - Failing
+
+    func testFailsWhenNoNotificationsArePosted() {
+        let testNotification = NSNotification(name: "Foo", object: nil)
+        let pattern = try! "expected to equal <[NSConcreteNotification %addr% {name = Foo}]>, got <[]>"
```
Comment (jeffh) [inline NimbleTests/Matchers/PostNotificationTest.swift]: I think a the need for regular expression failure messages can be removed by using NSString's format to get the memory address. [beIdenticalTo's tests](https://github.com/Quick/Nimble/blob/master/NimbleTests/Matchers/BeIdenticalToTest.swift#L17) do this:

```swift
func testBeIdenticalToPositiveMessage() {
    let num1 = NSNumber(integer:1)
    let num2 = NSNumber(integer:2)
    let message = NSString(format: "expected to be identical to <%p>, got <%p>", num2, num1)
    failsWithErrorMessage(message.description) {
        expect(num1).to(beIdenticalTo(num2))
    }
}
```
I don't think there's a need to add regular expression matching for `failsWithErrorMessage`.

File: Nimble/Utils/NotificationCollector.swift [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+//  NotificationCollector.swift
+//  Nimble
+//
+//  Created by Brian Gerstle on 12/31/15.
+//  Copyright © 2015 Jeff Hui. All rights reserved.
+//
+
+import Foundation
+
+class NotificationCollector {
```
Comment (jeffh) [inline Nimble/Utils/NotificationCollector.swift]: Nitpick: Please be explicit about visibility scope for top-level types (add `internal` keyword).

**Comment by jeffh** [comment]
Hey @bgerstle,

Thanks for the PR. I like what you arrived at as a solution. It's simplier than having to write custom matcher for all permutations.

I've written some feedback inlined to the PR. Please reply on this PR when you want me to review it again since github doesn't notify me when a PR changes. :+1:

File: Nimble/Utils/NotificationCollector.swift (hunk 2 above)
Comment (bgerstle) [inline Nimble/Utils/NotificationCollector.swift]: > I think the postedNotifications func is causing problems that forces expect(expression: ...) form you're talking about. postedNotification should return the observed notifications:

You're probably right, didn't think about how `@autoclosure` could be affecting parameter type inference.  I'll try your suggestion.

**Comment by bgerstle** [comment]
@jeffh I've re-re-refactored the API in a way that uses more conventional Nimble syntax while also removing the need for `expression:` label.  I've also addressed your other comments

**Comment by bgerstle** [comment]
@jeffh If you like this new approach, I can add ObjC API & update README

**Comment by bgerstle** [comment]
@jeffh looks like there's a [flaky ObjC test](https://travis-ci.org/Quick/Nimble/jobs/100944007#L2216), probably unrelated to my changes.

**Comment by jeffh** [comment]
Yeah, that’s related to https://github.com/Quick/Nimble/pull/224 <https://github.com/Quick/Nimble/pull/224>, so don’t worry about that.

I kind of like original helper function that captured the notifications in closure. It seems like it would be a bit harder this matcher to compare notifications by notification name, for example.

But I personally don’t use NSNotifications that much, so I’m not drawing from a lot of experience testing them.

- Jeff

**Comment by bgerstle** [comment]
@jeffh I agree that matching the name only is probably the most common use case and having the helper function allows users to map the output and compare only the field(s) they want.  IMHO it should be in the matcher and not in a helper for two reasons:

- I didn't like the helper function since it didn't match the usage of anything else in Nimble
- Keeping all the matching logic together results in better readability

For example:

``` swift
expect { ... }.to(postNotifications(contains(notificationWithName("Foo")))
```

IMHO this is more readable than:

``` swift
expect { postedNotifications { ... }.map { $0.name } }.to(equal("Foo"))
```

and allows users to mix and match concerns for individual notifications:

``` swift
expect { ... }.to(postNotifications(contains(notificationWithName("Foo"),  testNotificationWithSpecificNameAndObject))
```

Also, I would postulate these kinds of field-specific matching use cases are common enough to justify their own generic matcher.  For example, OCHamcrest implements `hasProperty` so you can assert that an object's property value satisfies a matcher.  For example: `assertThat(obj, hasProperty(@"foo", startsWith(@"bar"))`.  This is obviously harder to do with Swift (especially since `var` isn't accessible via getter function), but would make more sense if a functional approach were taken (e.g. lenses).

**Comment by jeffh** [comment]
This looks good, although the work on Linux support is conflicting with the PR, I'll see if I can get a merge in after that all settles.

**Comment by bgerstle** [comment]
thanks @jeffh i'll merge in master soon and address remaining checklist items. would like to see this in as well :grinning:

**Comment by mgetzbw** [comment]
Will this support `.PostASAP` and `.PostWhenIdle`?

**Comment by bgerstle** [comment]
@mgetzbw I believe so, using `toEventually(postNotification(...))`, but I'll have to check. Thanks for bringing it up!

**Comment by mgetzbw** [comment]
@bgerstle I bring it up because I've had trouble getting those two to work properly even in XCTest so I'm curious if you'd found a solution. `.PostWhenIdle` is probably going to be the hardest because it requires the runloop to be "Idle" for a given definition of idle.

**Comment by bgerstle** [comment]
@mgetzbw like I said, this should work as long as the notification is "eventually" posted to the observed center.  we might not be able to test `.PostWhenIdle` in practice because XCTest/Nimble will be polling to check if the matcher was satisfied. Might need to adjust the polling interval, but either way I'll report back with whether or not I've tested `.PostWhenIdle` specifically

**Comment by mgetzbw** [comment]
@bgerstle it strikes me that `.PostWhenIdle` may not be testable, as thread waits prevent the runloop from running, and the docs on `.PostWhenIdle` specifically mention that it runs as the last thing in the queue. This would preclude being able to race to the end unless we are able to run the test on an alternate runloop, which itself creates innumerable problems.

**Comment by bgerstle** [comment]
@mgetzbw perhaps, but runloop pumping might be sufficient to simulate idleness, which is [what Nimble does while waiting for async expectations](https://github.com/Quick/Nimble/blob/master/Sources/Nimble/Utils/Async.swift#L313)

**Comment by jeffh** [comment]
Hey @bgerstle,

I've merged in the changes and resolved the conflicts, there's still missing the README and Objective-C support. Are you interested in picking that up?

**Comment by bgerstle** [comment]
Yeah! I'll try to get that in soon.

**Comment by jeffh** [comment]
Hmm, the isEventually postNotifications test seems flaky: https://circleci.com/gh/jeffh/Nimble/25

**Comment by bgerstle** [comment]
Could that be at all related to the stall on main thread issues?

**Comment by jeffh** [comment]
Don't think so, it looks like other notifications fired during that time:

```swift
expected to eventually equal <[NSConcreteNotification 0x7fb0a24df940 {name = Foo}]>,
got <[__CFNotification 0x7fb0a24e1b00 {name = NSUserDefaultsDidChangeNotification; object = <NSUserDefaults: 0x7fb0a2400f10>},
__CFNotification 0x7fb0a24e1b80 {name = com.apple.CFPreferences._didChange; object = com.apple.CFPreferences; userInfo = {}},
__CFNotification 0x7fb0a24e1cb0 {name = NSUserDefaultsDidChangeNotification; object = <NSUserDefaults: 0x7fb0a2400f10>},
__CFNotification 0x7fb0a24e1ce0 {name = com.apple.CFPreferences._didChange; object = com.apple.CFPreferences; userInfo = {}},
__CFNotification 0x7fb0a24e1bb0 {name = NSUserDefaultsDidChangeNotification; object = <NSUserDefaults: 0x7fb0a2400f10>},
__CFNotification 0x7fb0a24c9b00 {name = com.apple.CFPreferences._didChange; object = com.apple.CFPreferences; userInfo = {}},
__CFNotification 0x7fb0a24c9b90 {name = NSUserDefaultsDidChangeNotification; object = <NSUserDefaults: 0x7fb0a2400f10>},
__CFNotification 0x7fb0a24c9bc0 {name = com.apple.CFPreferences._didChange; object = com.apple.CFPreferences; userInfo = {}},
__CFNotification 0x7fb0a24c9bf0 {name = NSUserDefaultsDidChangeNotification; object = <NSUserDefaults: 0x7fb0a2400f10>},
__CFNotification 0x7fb0a24c9c20 {name = com.apple.CFPreferences._didChange; object = com.apple.CFPreferences; userInfo = {}},
NSConcreteNotification 0x7fb0a24df940 {name = Foo}]>
```

**Comment by bgerstle** [comment]
Ah, that's a relief. Can either change expectation to "contains" or use a
private Notification Center.

**Comment by jeffh** [comment]
Using a private one probably wouldn't hurt.

-- 
Jeff Hui

**Pull request merged**